fuels-signers/src/wallet/mod.rs

```diff
@@ -83,8 +100,144 @@ impl Wallet {
+    /// Previous witnesses in this `Transaction` are preserved, allowing
+    /// the caller to implement multisig (transactions with many signatures).
+    pub fn append_signatures(&self, tx: &mut Transaction, sigs: &[Signature]) {
+        let new_witnesses: Vec<Witness> = sigs
+            .iter()
+            .map(|sig| Witness::from(sig.compact.as_ref()))
+            .collect();
+
+        let mut witnesses: Vec<Witness> = tx.witnesses().to_vec();
+        match witnesses.len() {
+            0 => tx.set_witnesses(new_witnesses),
+            _ => {
+                witnesses.extend(new_witnesses);
+                tx.set_witnesses(witnesses)
+            }
+        }
+    }
+
+    /// Transfer funds from this wallet to another `Address`.
+    /// Fails if amount for color is larger than address's spendable coins.
+    ///
+    /// # Examples
+    /// ```
+    /// use fuels_signers::provider::Provider;
+    /// use fuels_signers::{LocalWallet, Signer};
+    /// use fuels_signers::util::test_helpers::{
+    ///     setup_address_and_coins, setup_local_node, setup_test_provider,
+    /// };
+    /// use fuel_tx::{Bytes32, Color, Input, Output, UtxoId};
+    /// use rand::{rngs::StdRng, RngCore, SeedableRng};
+    /// use secp256k1::SecretKey;
+    /// use std::str::FromStr;
+    ///
+    /// async fn foo() -> Result<(), Box<dyn std::error::Error>> {
+    ///   // Setup test wallets with 1 coin each
+    ///   let (pk_1, mut coins_1) = setup_address_and_coins(1, 1);
+    ///   let (pk_2, coins_2) = setup_address_and_coins(1, 1);
+    ///   coins_1.extend(coins_2);
+    ///
+    ///   // Setup a provider and node with both set of coins
+    ///   let provider = setup_test_provider(coins_1).await;
+    ///
+    ///   // Create the actual wallets/signers
+    ///   let wallet_1 = LocalWallet::new_from_private_key(pk_1, provider.clone()).unwrap();
+    ///   let wallet_2 = LocalWallet::new_from_private_key(pk_2, provider).unwrap();
+    ///
+    ///   // Transfer 1 from wallet 1 to wallet 2
+    ///   let _receipts = wallet_1
+    ///        .transfer(&wallet_2.address(), 1, Default::default())
+    ///        .await
+    ///        .unwrap();
+    ///
+    ///   let wallet_2_final_coins = wallet_2.get_coins().await.unwrap();
+    ///
+    ///   // Check that wallet two now has two coins
+    ///   assert_eq!(wallet_2_final_coins.len(), 2);
+    ///   Ok(())
+    /// }
+    /// ```
+    pub async fn transfer(
+        &self,
+        to: &Address,
+        amount: u64,
+        color: Color,
+    ) -> io::Result<Vec<Receipt>> {
+        let spendable = self.get_spendable_coins(&color, amount).await?;
+
+        let mut inputs: Vec<Input> = vec![];
+        let mut outputs: Vec<Output> = vec![];
+
+        let mut amount_to_send = 0;
+
+        let output_coin = Output::coin(*to, amount, color);
+        outputs.push(output_coin);
+
+        for coin in spendable {
+            if amount_to_send >= amount {
+                break;
```

Comment:
This is counter productive to the random-improve algorithm. The intention is that the use should spend **_all_** the coins returned by the API, as the complete combination of coins should reduce dust in the future.

Reply:
For more details, check out the algorithm docs here: https://input-output-hk.github.io/cardano-coin-selection/haddock/cardano-coin-selection-1.0.1/Cardano-CoinSelection-Algorithm-RandomImprove.html

Basically we want the input to be on avg 2*amount, that way the change output will be approximately the right size for the next transaction.

Reply:
Right, so the API is returning the exact amount of coins to be spent (or error). But do we still have to create one `Input::Coin` per spendable coin? So we would still keep the loop to create the inputs but remove the `break;`? i.e.

```rust
for coin in spendable {
    let input_coin = Input::coin(
        UtxoId::from(coin.utxo_id),
        coin.owner.into(),
        coin.amount.0,
        color,
        0,
        0,
        vec![],
        vec![],
    );

    inputs.push(input_coin);

    amount_to_send += coin.amount.0;
}

// Calculate change, if applicable
if amount_to_send > amount {
    let change = Output::change(self.address(), amount_to_send - amount, color);
    outputs.push(change);
}
```

Reply:
Yep that's correct, although you don't need to calculate the change here since the block producer will do that automatically based on the gasPrice vs gasUsed etc...

Reply:
You still need to set who the owner of the change will be, but the amount is automatically calculated for every change output based on the color and freebalances.

Reply:
So that logic above can become just this?

```rust
for coin in spendable {
    let input_coin = Input::coin(
        UtxoId::from(coin.utxo_id),
        coin.owner.into(),
        coin.amount.0,
        color,
        0,
        0,
        vec![],
        vec![],
    );

    inputs.push(input_coin);
}

outputs.push(Output::change(self.address(), 0, color));
```

Reply:
yep, exactly!

Reply:
Neat!

Reply:
I think this release of the node your using should have correct change amounts already, but the unreleased version on master is more complete.

Reply:
Yes!! I just noticed (last commit here) that the test for sending + change is still working even without calculating the change here. 😄

Reply:
Nice!